source/common/upstream/upstream_impl.h

```diff
@@ -265,6 +269,7 @@ class HostImpl : public HostDescriptionImpl,
   static Network::ClientConnectionPtr
   createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& cluster,
                    const Network::Address::InstanceConstSharedPtr& address,
+                   const std::vector<Network::Address::InstanceConstSharedPtr>& address_list,
```

Comment:
So this is a bit of a work in progress. We have another piece of work to actually populate this in production and I suspect we'll end up tweaking the API here, but I'm not sure. I'd be inclined to punt the consolidate to the follow up PR.

That being said, the idea here is that traditionally Envoy creates a connection to a single IP, of course. The Envoy HostImpl object has a single address which is used when creating a connection. This address comes from the host resolution which happened before the HostImpl was created. We're going to change this so that host resolution optionally ends up providing the full list of addresses to the HostImpl. We'd also like to run v4 (A records) and v6 (AAAA records) resolution in parallel and delivering the results as they arrive. This seems likely to require a tweak to the API involved, but we'll get to that when we get there :)

Ok, so to your specific question... Today a host always has an address and sometimes has an addressList (in the tests which call setAddressList). We *could* have this method simply take an address list and treat the first entry as the main address. The downside is that this would require an extra vector allocation in the case where the HostImpl itself has no address list. I don't love this idea but it's not the end of the world. Happy to do it if you think it's best?

Reply:
It sounds a bit hairy so let's save that until we want to properly support a list of addresses, thanks for explaining!